.changeset/moody-rocks-accept.md

```diff
@@ -0,0 +1,28 @@
+---
+'@backstage/plugin-catalog': patch
+'@backstage/plugin-catalog-react': patch
+---
+
+The `useEntityList` hook has been updated to send a list of entity fields to the
+backend. This optimizes the catalog query, especially in the case of a catalog
+with many API definitions.
+
+This is a transparent change _unless_ you have added custom components that use
+the `useEntityList` hook. If this applies to you, you can update those
+components to use the new exported `addEntityFields` function to add any entity
+fields used by your component:
+
+```diff
+export function MyCustomPicker = () => {
+-    const { updateFilters, filters } = useEntityList();
++    const { updateFilters, filters, addEntityFields } = useEntityList();
+    const [values, setValues] = useState<string[]>([]);
+
++    useEffect(() => addEntityFields(['spec.otherfield']), [addEntityFields]);
```

Comment:
This could just as well have been `useEntityList({ additionalEntityFields: [] })` or similar right? Analogous to the initial value given to `useState` and similar. Could still have the method as suggested in the PR, for things that wish to modify the set deep down (maybe in custom columns or something?)

Reply:
I like the signature of that, but I'm not sure it actually works with the React Context underneath 🤔  since context consumers can't provide initialization values. The hook itself could call `addEntityFields` with the argument, but I think it would be called repeatedly on each component render. The function is idempotent, but I don't like relying on that with wasteful calls.

Reply:
You'd probably call useEffect with the addition inside useEntityList - that would work right? I'm mainly after alleviating the need for that whole dance all over the place.

Reply:
The useEffect would have `additionalEntityFields` as a dependency though, which has no guarantee of being stable (in fact, most certainly passed in-line or through a calculated variable).

Reply:
Doesn't have to. This is a fine moment to break that. Either you go ugly and lint-ignore it, or you go the verbose route of stashing it in a useRef (which is one more useEffect just to keep the ref needlessly fresh).

Reply:
Agree that it should be possible to get working while making it easy for the caller. Tbh I'm not sure about the clarity though, reading `useEntityList({ additionalEntityFields: [] })` I'd assume that it's a local modification on the consumer end rather than mutating the context value. I feel the current design is a bit clearer in that it's sending things back up the context. Might just be a naming issue though.

Either way, some more ideas further down that stirs this up quite a bit

Reply:
isn't this a breaking change considering that things will break in case people are using custom filters?